atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/PreStartHandlingTransactionService.java

```diff
@@ -0,0 +1,84 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.service;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import javax.annotation.CheckForNull;
+
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+
+/**
+ * This service handles queries for timestamps before {@link com.palantir.atlasdb.AtlasDbConstants#STARTING_TS}
+ * as follows:
+ *
+ * - Gets of timestamps before {@link com.palantir.atlasdb.AtlasDbConstants#STARTING_TS} will return
+ *   {@link com.palantir.atlasdb.AtlasDbConstants#STARTING_TS - 1}; in an AtlasDB context these correspond to
+ *   deletion sentinels that are written non-transactionally and thus always committed.
+ * - putUnlessExists to timestamps before {@link com.palantir.atlasdb.AtlasDbConstants#STARTING_TS} will throw an
+ *   exception.
+ *
+ * Queries for legitimate timestamps are routed to the delegate.
+ */
+public class PreStartHandlingTransactionService implements TransactionService {
+    private final TransactionService delegate;
+
+    public PreStartHandlingTransactionService(TransactionService delegate) {
+        this.delegate = delegate;
+    }
+
+    @CheckForNull
+    @Override
+    public Long get(long startTimestamp) {
+        if (!isTimestampValid(startTimestamp)) {
+            return AtlasDbConstants.STARTING_TS - 1;
+        }
+        return delegate.get(startTimestamp);
+    }
+
+    @Override
+    public Map<Long, Long> get(Iterable<Long> startTimestamps) {
+        Map<Boolean, List<Long>> classifiedTimestamps = StreamSupport.stream(startTimestamps.spliterator(), false)
+                .collect(Collectors.partitioningBy(PreStartHandlingTransactionService::isTimestampValid));
+
+        Map<Long, Long> result = Maps.newHashMap();
+        result.putAll(delegate.get(classifiedTimestamps.get(true)));
+        result.putAll(Maps.uniqueIndex(classifiedTimestamps.get(false), unused -> AtlasDbConstants.STARTING_TS - 1));
+        return result;
+    }
+
+    @Override
+    public void putUnlessExists(long startTimestamp, long commitTimestamp) throws KeyAlreadyExistsException {
+        if (!isTimestampValid(startTimestamp)) {
+            throw new SafeIllegalStateException("Attempted to putUnlessExists({}, {}) which is unexpected.",
```

Comment:
'disallowed' over 'unexpected' to be clearer that this is a bad thing to do, not just a case we didn't consider

Reply:
:+1: changed to disallowed